### DISCUSSION TITLE
HelpSystem 2.0

### DISCUSSION THREAD

**Discussion by SteveL-MSFT** [PowerShell/PowerShell#21196] [Enhancements]
There has been thoughts of a new HelpSystem since PowerShell Core 6.  Some items in mind:

- Native markdown support instead of MAML https://github.com/PowerShell/PowerShell/issues/3954
- Side-by-side support of versions of help aligned w/ versions of PS7 https://github.com/PowerShell/PowerShell/issues/1686
- On-demand download of help instead of full initial download of all help (even if never viewed)
- Fuzzy searching of help

Other ideas/requests for how help in PowerShell can be improved?

**Comment by MartinGC94** [comment]
I don't know if it should be considered a part of the help system, but it would be really nice if we could view the .NET type and method descriptions from inside the shell.  
Maybe `Update-Help` could download the same XML files that Visual Studio uses to display the help? They seem to be hosted in this repository, example: https://github.com/dotnet/dotnet-api-docs/blob/main/xml/System.Security.Principal/NTAccount.xml and PowerShell modules that include their types could either bundle the help in the module, or the module manifest could include a download link for the XML documentation.

**Reply by ThomasNieto** [reply]
Yes, this is one of my main asks as well for the help system to work with .NET types.

**Comment by ThomasNieto** [comment]
* Down level support to 5.1+
* Legacy MAML support
* Dynamic parameter support (PSProviders and other provider like systems)
  * #19301 
  * #19300
  * #13823 
* Localization support and PlatyPS headers in other languages
* Easier updatable help (not require a website and can leverage GitHub)
* Different help document types like FAQ, etc (HelpCategory class)
* Help structured data to be able to interact via cmdlet/script.
* Configurable current user and all users help file locations

**Comment by rhubarb-geek-nz** [comment]
> Other ideas/requests for how help in PowerShell can be improved?

As a developer I don't use the PowerShell command line help. Reading help using a console command line is the option of last resort.

* The help system should be external to the core PowerShell scripting engine and optional so I don't have to have it on a minimal system
* help for functions and cmdlets should be visible directly from PowerShell gallery, if I click on a cmdlet or function in the gallery it takes back to a high level of the gallery
* help should be an external package from a binary module in order to provide multi-language help
* don't replicate functionality that is already done better by a web-search or through Visual Code.

**Comment by kasini3000** [comment]
Until now, I am still using the "Microsoft PowerShell 2.0 Official Chinese Manual. chm" to send to beginners, which is located in Chinese in Win7.
I hope to have the "Powershell 7.4 Chinese Help Manual. chm" in Windows 11.

**Comment by 237dmitry** [comment]
Markdown would be a good format for help pages, the pager `less` renders the output of `Show-Markdown` properly. Alternatively, this could be the `troff` format of man pages.

**Comment by NoralK** [comment]
Here are my thoughts, coming from a PS 5.1 world updating to PS 7.4 while maintaining backward capabilities:

 - Better support of Dynamic Parameters
 - Output to this new format or "classic" PowerShell help
 - Support for basic formatting, like **bold**, _italics_, <ins>underline</ins>, and color ... I know there is not specification for underline and color but it ${\color{red}\text{is} \space \color{lightblue}\text{the} \space \color{orange}\text{21st}}$ ${\color{yellow}\text{century}}$ ... not timesharing on a 60s VAX machine with a dumb terminal over a 75-baud modem :-)
 - **Tags** for ${\color{red}\text{Breaking Changes}}$, ${\color{orange}\text{Deprecated}}$, ${\color{yellow}\text{Obsolete}}$, ${\color{cyan}\text{Experimental}}$, or self-defined Tags like ${\color{magenta}\text{UseAtOwnRisk}}$ and, as shown, it would be nice to be able to colorize these tags as well
 - The version of the module or cmdlet **outputted** to support multiple versions, like for PNP.PowerShell - one version for PS-ISE but another for PS 7
 - Markdown seems the way to go for structuring these files, it is simple and widely supported. The downside is there is many different ways to markdown a file, like using the variants from the GitHub or Markdown Extra ... the question would be what would and would not be valid for "PowerShell Help Markdown"
- VS Code PowerShell extension updated to better handle help - all commands in the left pane is not helpful and the help in PS window not helpful.  Something like opening help a new file or a popup that can be made modal, but that is another group :-)
- Output the help to HTML with a syntaxhighlighter with a tree structure like [PNP.PowerShell document pages](https://pnp.github.io/powershell/cmdlets/index.html), but for multiple modules ... think current PS-ISE for displaying help with the main window showing the help a command.  Cmdlets like `Export-HelpByModule -module myModule -format HTML` or `Export-HelpByTopic -module myModule -topic New-Item -format HTML`
- Reaching here, but it would be nice to display an image, if the help viewer supported it.  It would be nice to have a dedicated MS Help Viewer ... Oh, WAIT a sec, Visual Studio does have free one and it uses HTML but I bet a new Help View, say v3.0 or better yet open-source, could support markdown as well

These are wishes that are in order of my priority, but if I had to pick it would be the ability to Tag a help topic.

**Comment by MatejKafka** [comment]
I don't like having separate XML help files. There are 3rd party solutions for that (https://www.nuget.org/packages/XmlDoc2CmdletDoc), but I'd love to have docstring support as a first-party, well-supported feature from PowerShell, preferably even without extra build steps, if the runtime performance penalty wouldn't be too high.
